src/intrinsics.cpp

```diff
@@ -0,0 +1,558 @@
+    } else if (t == LLVMTypes::FloatVectorPointerType) {
+        return PointerType::GetUniform(AtomicType::VaryingFloat);
+    } else if (t == LLVMTypes::DoubleVectorPointerType) {
+        return PointerType::GetUniform(AtomicType::VaryingDouble);
+    }
+
+    return nullptr;
+}
+
+/** Create ISPC symbol for LLVM intrinsics and add it to the given module.
+
+    @param func            llvm::Function for the intrinsic to be added
+    @param symbolTable     SymbolTable in which to add symbol definitions
+    @return                Symbol created for the LLVM::Function
+ */
+Symbol *lCreateISPCSymbolForLLVMIntrinsic(llvm::Function *func, SymbolTable *symbolTable) {
+    Symbol *existingSym = symbolTable->LookupIntrinsics(func);
+    if (existingSym != nullptr) {
+        return existingSym;
+    }
+    SourcePos noPos;
+    noPos.name = "LLVM Intrinsic";
+    const llvm::FunctionType *ftype = func->getFunctionType();
+    std::string name = std::string(func->getName());
+    const Type *returnType = lLLVMTypeToISPCType(ftype->getReturnType(), false);
+    if (returnType == nullptr) {
+        Error(SourcePos(),
+              "Return type not representable for "
+              "Intrinsic %s.",
+              name.c_str());
+        // return type not representable in ispc -> not callable from ispc
+        return nullptr;
+    }
+    llvm::SmallVector<const Type *, 8> argTypes;
+    for (unsigned int j = 0; j < ftype->getNumParams(); ++j) {
+        const llvm::Type *llvmArgType = ftype->getParamType(j);
+        const Type *type = lLLVMTypeToISPCType(llvmArgType, false);
+        if (type == nullptr) {
+            Error(SourcePos(),
+                  "Type of parameter %d not "
+                  "representable for Intrinsic %s",
+                  j, name.c_str());
+            return nullptr;
+        }
+        argTypes.push_back(type);
+    }
+    FunctionType *funcType = new FunctionType(returnType, argTypes, noPos);
+    Debug(noPos, "Created Intrinsic symbol \"%s\" [%s]\n", name.c_str(), funcType->GetString().c_str());
+    Symbol *sym = new Symbol(name, noPos, Symbol::SymbolKind::Function, funcType);
+    sym->function = func;
+    symbolTable->AddIntrinsics(sym);
+    return sym;
+}
+
+static ISPCIntrinsics lLookupISPCInstrinsic(const std::string &name) {
+    llvm::StringRef ref(name);
+    if (ref.starts_with("llvm.ispc.atomicrmw.")) {
+        return ISPCIntrinsics::atomicrmw;
+    }
+    if (ref.starts_with("llvm.ispc.cmpxchg.")) {
+        return ISPCIntrinsics::atomicrmw;
+    }
+    if (ref.starts_with("llvm.ispc.fence.")) {
+        return ISPCIntrinsics::fence;
+    }
+    for (unsigned i = 1; i <= (unsigned)ISPCIntrinsics::stream_store; i++) {
+        if (name == ISPCIntrinsicsNames[i]) {
+            return (ISPCIntrinsics)i;
+        }
+    }
+    return ISPCIntrinsics::not_intrinsic;
+}
+
+static llvm::VectorType *lGetDoubleVectorType(llvm::Type *type) {
```

Comment:
I think the better name would be `lGetDoubleWidthVectorType`

Reply:
Done